pyscf/pbc/df/df.py

```diff
@@ -831,6 +831,67 @@ def shape(self):
         else: # For mpi4pyscf, pyscf-1.5.1 or older
             return dat.shape
 
+def _hstack_datasets(data_to_stack, slices=None):
+    """Faster version of numpy.hstack for h5py datasets
+
+    Parameters
+    ----------
+    data_to_stack : list of h5py.Dataset or np.ndarray
+        Datasets/arrays to be stacked along first axis
+    slices: tuple of slices, or just one slice, along the lines of numpy.s_.
+
+    Returns
+    -------
+    numpy.ndarray
+        The stacked data, equal to numpy.hstack([dset[slices] for dset in data_to_stack])
+    """
+    # Step 1. Calculate the shape of the output array, and store it
+    # in res_shape.
+    res_shape = list(data_to_stack[0].shape)
+    dset_shapes = [x.shape for x in data_to_stack]
+
+    if not (isinstance(slices, tuple) or isinstance(slices, list)):
+        # If slices is not a tuple, we assume it is a single slice acting on axis 0 only.
+        slices = (slices,)
+
+    for i, t in enumerate(slices):
+        ts = numpy.s_[t]
+        if i == 1:
+            ax1widths_sliced = [len(range(sh[1])[ts]) for sh in dset_shapes]
+        else:
+            # Except along axis 1, we assume the dimensions of all datasets are the same.
+            # If they aren't, an error gets raised later.
+            res_shape[i] = len(range(res_shape[i])[ts])
+    if len(slices) <= 1:
+        ax1widths_sliced = [sh[1] for sh in dset_shapes]
+
+    # Final dim along axis 1 is the sum of the post-slice axis 1 widths.
+    res_shape[1] = sum(ax1widths_sliced)
+
+    # Step 2. Allocate the output buffer
+    out = numpy.empty(res_shape, dtype=numpy.result_type(*data_to_stack))
```

Comment:
Here is an implicit type conversion.  `np.result_type` requires the arguments to be arrays or dtypes. The h5dset may or may not be compatible with arrays. It's better to first extract data types for each element in `data_to_stack` then pass to `np.result_type()` .

Reply:
Done!